File: backend/src/constants/combat.ts
/**
 * Combat mechanic constants — hit, crit, and ailment formulas.
 * Update when LE patches change ailment ratios, durations, or crit caps.
 */

/** Player crit chance hard cap */
export const CRIT_CHANCE_CAP = 0.95 as const;

/** BuildStats defaults */
export const BASE_CRIT_CHANCE = 0.05 as const;
export const BASE_CRIT_MULTIPLIER = 1.5 as const;

/** Ailment DoT — Last Epoch baseline durations per stack.
 * VERIFIED: 1.4.3 spec §4 — each ailment has flat base DPS per stack (see
 * BLEED_BASE_DPS / IGNITE_BASE_DPS / POISON_BASE_DPS in the Python constants).
 * The old legacy ratio constants (BLEED_BASE_RATIO etc.) were incorrect and
 * have been removed.
 */
export const BLEED_DURATION = 4.0 as const;      // seconds
export const IGNITE_DURATION = 3.0 as const;     // seconds
export const POISON_DURATION = 3.0 as const;     // seconds

File: backend/app/constants/combat.py
"""
Combat mechanic constants — hit, crit, and ailment formulas.

Update when LE patches change ailment ratios, durations, or crit caps.
"""

# Player crit chance hard cap (100% — can always reach guaranteed crits)
CRIT_CHANCE_CAP: float = 1.0

# BuildStats defaults — also used as dataclass field defaults in stat_engine.py
BASE_CRIT_CHANCE: float = 0.05
# VERIFIED: 1.4.3 spec §2.2 — base crit multiplier is 200% (2.0×), not 150%
BASE_CRIT_MULTIPLIER: float = 2.0

# Hit damage variance: ±25% for hits, none for DoTs
# VERIFIED: 1.4.3 spec §2.1 — hit damage rolls uniformly in [0.75×, 1.25×]
HIT_DAMAGE_VARIANCE: float = 0.25

# Ailment DoT — Last Epoch flat base damage per stack, pre-modifier
# These are the base DPS values per stack; total = stacks × base × (1 + increased)
IGNITE_BASE_DPS: float = 40.0     # 40 fire DoT per second per stack
IGNITE_DURATION: float = 3.0      # seconds

# UNVERIFIED: Last Epoch 1.4.3 publishes no single authoritative "base bleed DPS"
# number. 43.0 is a community-derived approximation used across the engine and
# regression tests; re-check before any patch-driven recalibration.
BLEED_BASE_DPS: float = 43.0      # ~43 physical DoT per second per stack
BLEED_DURATION: float = 4.0       # seconds

POISON_BASE_DPS: float = 28.0     # 28 poison DoT per second per stack
POISON_DURATION: float = 3.0      # seconds

ELECTRIFY_BASE_DPS: float = 44.0  # 44 lightning DoT per second per stack
ELECTRIFY_DURATION: float = 2.5   # seconds

# Boss ailment effectiveness reduction (60% less effective vs bosses)
BOSS_AILMENT_REDUCTION: float = 0.60
